### PR TITLE
Bq date partitioning

### PR DESCRIPTION
**Pull request by drewbanin**
This branch adds support for date partitioning on BigQuery in dbt.

TODO:
 - [x] fix tests
 - [x] prevent views from overwriting date-partitioned table, as this could delete a significant amount of data!

### Run for a single day, hard-coded

Simple usage, specify a single date manually:

```sql
{{
    config(
        materialized='table',
        partition_date='20180101',
    )
}}

select *
from `public`.`events_20180101`
```

### Run for a range of days

More complex usage, specify a range of dates. The `date_sharded_table` macro will interpolate the 8-digit date for each of the days between January 1st and January 10th, inclusive. In this way, the resulting date partitioned table will have 10 partitions, one for each day, built from the corresponding date-sharded `events_[YYYYMMDD]` tables.

```sql
{{
    config(
        materialized='table',
        partition_date='20180101,20180110',
    )
}}

select *
from `public`.`{{ date_sharded_table('events_') }}`
```

### Dynamically specify partition date(s)

Use a variable instead of hardcoding a single date. The variable defaults to "yesterday" if  `partition_date` is not provided.

```sql
-- macros/datetime.sql
{% macro yesterday() -%}
    {%- set delta = modules.datetime.timedelta(days=-1) -%}
    {{ return(run_started_at + delta).strftime('%Y%m%d') }}
{%- endmacro %}
```

```sql
--models/partitioned.sql
{{
    config(
        materialized='table',
        partition_date=var('partition_date', yesterday()),
    )
}}

select *
from `public`.`{{ date_sharded_table('events_') }}`
```

This branch is intended to be used in conjunction with https://github.com/fishtown-analytics/dbt/pull/640 to supply variables to date partitioned tables on the command line.

### Additional configuration

A full list of configuration options for date partitioned tables is shown below:

```sql
{{
    config(
        materialized='table',
        partition_date='2018-01-01,2018-01-10,
        partition_date_format='%Y-%m-%d',
        verbose=True
    )
}}
```

- `partition_date_format` : The date format (using [strptime/stftime](https://docs.python.org/2/library/datetime.html#strftime-and-strptime-behavior) conventions) with which to parse the `partition_date` field. Default: `%Y%m%d`
- `verbose` : If set to `True`, dbt will output one log line for each date partition created during the invocation of the date partitioned model. Default: `False`

```
$ dbt run
Found 1 models, 1 tests, 0 archives, 0 analyses, 49 macros, 0 operations

19:43:10 | Concurrency: 1 threads (target='dev')
19:43:10 |
19:43:18 | 6 of 6 START table model dbt_dbanin.partitioned.................. [RUN]
19:43:18 | -> Running for day 20180101
19:43:19 | -> Running for day 20180102
19:43:21 | -> Running for day 20180103
19:43:22 | 1 of 1 OK created table model dbt_dbanin.partitioned............. [CREATED 3 PARTITIONS in 4.60s]
19:43:22 |
19:43:22 | Finished running 1 table models in 13.23s.

Completed successfully

Done. PASS=1 ERROR=0 SKIP=0 TOTAL=1
```

### CODE REVIEW

**Comment by cmcarthur** [comment]
my gut reaction to the `20180101,20180110` syntax is that it's not ideal. it seems like it'd be better to pass a `range(20180101,20180110)` as the `partition_date`. that's more powerful & would let you do things like `partitiondate=[20180101, 20180201]` (rebuild jan 1 and feb 1, specifically). but my guess is that you used `20180101,20180110` so that you can pass in a variable range on the command line, is that right?

File: dbt/adapters/bigquery.py [hunk 1]
```diff
@@ -148,27 +148,37 @@ def query_for_existing(cls, profile, schemas, model_name=None):
...
             schemas = [schemas]
 
+        conn = cls.get_connection(profile, model_name)
+        client = conn.get('handle')
+
         all_tables = []
         for schema in schemas:
             dataset = cls.get_dataset(profile, schema, model_name)
-            all_tables.extend(dataset.list_tables())
+            all_tables.extend(client.list_tables(dataset))
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: oof, is this the API change you were referencing?

File: dbt/adapters/bigquery.py [hunk 2]
```diff
@@ -148,27 +148,37 @@ def query_for_existing(cls, profile, schemas, model_name=None):
...
     def drop(cls, profile, schema, relation, relation_type, model_name=None):
+        conn = cls.get_connection(profile, model_name)
+        client = conn.get('handle')
+
         dataset = cls.get_dataset(profile, schema, model_name)
         relation_object = dataset.table(relation)
-        relation_object.delete()
+        client.delete_table(relation_object)
+
+        cls.release_connection(profile, model_name)
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: i don't think it's correct to release the connection here -- what if you drop the table first, and then create it? i guess for bigquery it makes no difference, but better to exclude it if extraneous

File: dbt/adapters/bigquery.py [hunk 3]
```diff
@@ -244,55 +285,44 @@ def execute_model(cls, profile, model, materialization, model_name=None):
...
+        # this blocks until the query has completed
+        with cls.exception_handler(profile, 'create dataset', model_name):
+            iterator = query_job.result()
 
         res = []
         if fetch:
-            res = cls.fetch_query_results(query)
+            res = list(iterator)
+
+        cls.release_connection(profile, model_name)
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: same comment as on drop re: releasing connection

File: dbt/adapters/bigquery.py [hunk 4]
```diff
@@ -308,15 +338,24 @@ def add_begin_query(cls, profile, name):
...
-        dataset = cls.get_dataset(profile, schema, model_name)
+        conn = cls.get_connection(profile, model_name)
+        client = conn.get('handle')
 
+        dataset = cls.get_dataset(profile, schema, model_name)
         with cls.exception_handler(profile, 'create dataset', model_name):
-            dataset.create()
+            client.create_dataset(dataset)
+
+        cls.release_connection(profile, model_name)
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: ditto

File: dbt/adapters/bigquery.py [hunk 5]
```diff
@@ -308,15 +338,24 @@ def add_begin_query(cls, profile, name):
...
-        for table in dataset.list_tables():
-            table.delete()
+    def drop_tables_in_schema(cls, profile, dataset):
+        conn = cls.get_connection(profile)
+        client = conn.get('handle')
+
+        for table in client.list_tables(dataset):
+            client.delete_table(table.reference)
+
+        cls.release_connection(profile, name=None)
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: +

File: dbt/include/global_project/macros/materializations/bigquery.sql [hunk 6]
```diff
@@ -16,6 +25,80 @@
...
+  {% if day_count < 0 %}
+    {% set msg -%}
+        Partiton start date is after the end date ({{ start_date }}, {{ end_date }})
+    {%- endset %}
+    {{ exceptions.raise_compiler_error(msg, model) }}
+  {% endif %}
+
+  {% for i in range(0, day_count + 1) %}
+    {% set the_day = (modules.datetime.timedelta(days=i) + start_date).strftime('%Y%m%d') %}
+    {% if verbose %}
```
Comment (cmcarthur) [inline dbt/include/global_project/macros/materializations/bigquery.sql]: nice

File: dbt/include/global_project/macros/materializations/bigquery.sql [hunk 7]
```diff
@@ -16,6 +25,80 @@
...
+
+{% endmacro %}
+
+{% macro convert_datetime(date_str, date_fmt) %}
+
+  {% set error_msg -%}
+      The provided partition date '{{ date_str }}' does not match the expected format '{{ date_fmt }}'
+  {%- endset %}
+
+  {% set res = try_or_compiler_error(error_msg, modules.datetime.datetime.strptime, date_str.strip(), date_fmt) %}
```
Comment (cmcarthur) [inline dbt/include/global_project/macros/materializations/bigquery.sql]: this is really clever haha

File: test/integration/base.py [hunk 8]
```diff
@@ -173,6 +173,7 @@ def setUp(self):
 
         # it's important to use a different connection handle here so
         # we don't look into an incomplete transaction
+        adapter.cleanup_connections()
```
Comment (cmcarthur) [inline test/integration/base.py]: i think this obviates the need for release_connection everywhere else

File: dbt/adapters/bigquery.py (hunk 1 above)
Comment (drewbanin) [reply dbt/adapters/bigquery.py]: yeah :/

**Comment by drewbanin** [comment]
@cmcarthur my first cut of this used `start_date` and `end_date` as two different variables instead of a single `partition_date`. That worked pretty well, but it was confusing that when `end_date` isn't set, the date partitioning only runs for the `start_date`. Moreover, it's a little more difficult to type out a yaml dictionary with two elements on the command line IMO.

I like the idea of using `range` conceptually, but these dates are essentially strings, not integers! Eg:

```
range(20180131, 20180201)
```

^this would run for 20180131 (good) and then 20180132 (bad), so we'd need to implement our own sort of date_range function I think.

I think you're right though -- it's unusual that `partition_date` accepts a comma-separated string. One other option I can think of is to make the materialization accept a list of dates to run for, where this list can be generated by a macro from a start/end date pair. We can implement this macro in the global project to make this easy/transparent for users.

So:

```
$ dbt run --vars 'partition_date: "20180101, 20180131"'
```

Then in your model:
```
{{
    config(
        materialized='table',
        partition_date=date_range_to_list(var('partition_date')),
    )
}}

...
```

and then a macro which looks like:

```
{% macro date_range_to_list(range_str) %}

  start, end = range_str.split(",")
  dates = []
  for each date in range(start,end):
    dates.append(date)

  return dates

{% endmacro %}
```

So the CLI interface is the same, but the macro interface can work for a single date, a date range, or a smattering of random dates.

Let me know what you think about this kind of approach

**Comment by cmcarthur** [comment]
I like this approach a lot. We can write up docs on how to set your exact CLI syntax up for a project, but if someone really wanted to implement the other use case, they could do it themselves via:

```
{{config(partition_date(var('partition_dates').split(',')))}}
```

```
dbt run --vars 'partition_dates: 20180101, 20180201'
```

**Comment by drewbanin** [comment]
Ok, i'll do that

**Comment by drewbanin** [comment]
This is now implemented such that the `table` materialization can accept a list of "partitions". Eg:

```
{{
    config(
        materialized='table',
        partitions=['20180101', '20180102', '20180103'],
        verbose=True
    )
}}
```

The type checking is pretty loosey goosey here, so you can give a list of strings, list of ints, a single string, a single int, etc. These dates must be provided in BigQuery date format, ie. an 8-character series of digits.

To generate this list of dates, users can use the `dbt.partition_range` function built into the dbt global project. In practice, this looks like:

```
{{
    config(
        materialized='table',
        partitions=dbt.partition_range('20180101, 20180201'),
        verbose=True
    )
}}
```

This `partition_range` function will generate a list of dates in the range of the two provided dates. If only one date is provided, the resulting date range will only contain the date specified. This function also accepts an optional date format string. Finally, this macro can be combined with CLI vars to configure date ranges from the CLI, eg.

```
$ dbt run --model partitioned_model --vars 'dates: "20180101, 20180201"'
```

coupled with

```
{{
    config(
        materialized='table',
        partitions=dbt.partition_range(var('dates')),
        verbose=True
    )
}}
```

Users can further extend these macros to simplify patterns which they use frequently.

**Comment by cmcarthur** [comment]
lgtm!

**Pull request merged**